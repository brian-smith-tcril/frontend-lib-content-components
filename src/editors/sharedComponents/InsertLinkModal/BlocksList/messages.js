import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  blocksListSubsectionTitle: {
    id: 'blocks.list.subsection.title',
    defaultMessage: 'Subsections',
    description: 'Title for the subsections blocks',
  },
  blocksListUnitTitle: {
    id: 'blocks.list.unit.title',
    defaultMessage: 'Units',
    description: 'Title for the units blocks',
  },
});

export default messages;
