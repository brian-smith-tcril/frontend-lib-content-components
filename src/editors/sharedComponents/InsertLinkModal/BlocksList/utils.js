import cloneDeep from 'lodash.clonedeep';
import blockTypes from '../blockTypes';

/**
 * Retrieves a list of sections from the provided blocks object.
 *
 * @param {Object} blocks - The blocks object containing various block types.
 * @returns {Array} An array of section (type: chapter) blocks extracted from the blocks object.
 */
export const getSectionsList = (blocks = {}) => {
  const blocksList = Object.keys(blocks);
  return blocksList.reduce((previousBlocks, blockKey) => {
    const block = cloneDeep(blocks[blockKey]);
    if (block.type === blockTypes.section) {
      return [...previousBlocks, block];
    }

    return previousBlocks;
  }, []);
};

/**
 * Retrieves an array of child blocks based on the children list of a selected block.
 *
 * @param {Object} blockSelected - The selected block for which children are to be retrieved.
 * @param {Object} blocks - The blocks object containing various block types.
 * @returns {Array} An array of child blocks cloned from the blocks object.
 */
export const getChildrenFromList = (blockSelected, blocks) => {
  if (blockSelected.children) {
    return blockSelected.children.map((key) => cloneDeep(blocks[key]));
  }
  return [];
};
