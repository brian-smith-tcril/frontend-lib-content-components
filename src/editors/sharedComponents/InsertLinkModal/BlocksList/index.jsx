import { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, TransitionReplace, ActionRow } from '@openedx/paragon';
import { ArrowForwardIos, ArrowBack } from '@openedx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import blockTypes from '../blockTypes';
import { getSectionsList, getChildrenFromList } from './utils';

import messages from './messages';
import './index.scss';

const BlocksList = ({
  blocks,
  onBlockSelected,
  disableBlocks,
}) => {
  const intl = useIntl();
  const messageBlockType = {
    [blockTypes.section]: intl.formatMessage(
      messages.blocksListSubsectionTitle,
    ),
    [blockTypes.subsection]: intl.formatMessage(messages.blocksListUnitTitle),
    [blockTypes.unit]: intl.formatMessage(messages.blocksListUnitTitle),
  };

  const [blockState, setBlockState] = useState({
    blockSelected: {},
    type: blockTypes.subsection,
    hasNavigated: false,
    blocksNavigated: [],
  });

  const sections = getSectionsList(blocks);
  const subsections = getChildrenFromList(
    blockState.blockSelected,
    blocks,
  );
  const listItems = blockState.hasNavigated ? subsections : sections;

  const isBlockSelectedUnit = blockState.type === blockTypes.unit;
  const blockNameButtonClass = isBlockSelectedUnit ? 'col-12' : 'col-11';

  const handleSelectBlock = (block, navigate = false) => {
    if (navigate) {
      setBlockState({
        ...blockState,
        blocksNavigated: [...blockState.blocksNavigated, block.id],
        blockSelected: block,
        type: block.type,
        hasNavigated: true,
      });
    } else {
      onBlockSelected(block);
    }
  };

  const handleGoBack = () => {
    const newValue = blockState.blocksNavigated.filter(
      (id) => id !== blockState.blockSelected.id,
    );
    if (newValue.length) {
      const lastBlockIndex = newValue.length - 1;
      const blockId = newValue[lastBlockIndex];
      const newBlock = blocks[blockId];
      setBlockState({
        ...blockState,
        type: newBlock.type,
        hasNavigated: true,
        blockSelected: newBlock,
        blocksNavigated: newValue,
      });
    } else {
      setBlockState({
        ...blockState,
        type: blockState.section,
        hasNavigated: false,
        blockSelected: {},
      });
    }
  };

  return (
    <>
      {blockState.hasNavigated && (
        <ActionRow className="w-100 d-flex justify-content-space-between p-3">
          <Button
            variant="tertiary"
            className="col-1"
            onClick={handleGoBack}
            iconBefore={ArrowBack}
            data-testid="block-back-navigation"
          >
          &nbsp;
          </Button>
          <p className="col-11 text-center">{messageBlockType[blockState.type]}</p>
        </ActionRow>
      )}
      <div className="block-list-container">
        {listItems.map((block) => (
          <TransitionReplace key={`transition_${block.id}`}>
            <ActionRow
              key={block.id}
              className="w-100 d-flex justify-content-space-between p-3"
            >
              <Button
                variant="tertiary"
                className={`${blockNameButtonClass} py-4`}
                onClick={() => handleSelectBlock(block)}
                data-testid="block-name"
                disabled={disableBlocks}
              >
                <span className="w-100 text-left">{block.displayName}</span>
              </Button>
              {!isBlockSelectedUnit && (
                <Button
                  variant="tertiary"
                  className="col-1 py-4"
                  onClick={() => handleSelectBlock(block, true)}
                  data-testid="block-navigation"
                  iconAfter={ArrowForwardIos}
                  disabled={disableBlocks}
                >
                &nbsp;
                </Button>
              )}
            </ActionRow>
          </TransitionReplace>
        ))}
      </div>
    </>
  );
};

const blockShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  blockId: PropTypes.string.isRequired,
  lmsWebUrl: PropTypes.string.isRequired,
  legacyWebUrl: PropTypes.string.isRequired,
  studentViewUrl: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  displayName: PropTypes.string.isRequired,
  children: PropTypes.arrayOf(PropTypes.string),
});

BlocksList.defaultProps = {
  disableBlocks: false,
};

BlocksList.propTypes = {
  blocks: PropTypes.objectOf(blockShape).isRequired,
  onBlockSelected: PropTypes.func.isRequired,
  disableBlocks: PropTypes.bool,
};

export default BlocksList;
