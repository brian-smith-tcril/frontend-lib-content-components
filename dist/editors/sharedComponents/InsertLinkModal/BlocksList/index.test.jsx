import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import BlocksList from '.';

const mockBlocks = {
  'block-key': {
    id: 'block-key',
    blockId: 'edx_block-1',
    lmsWebUrl: 'http://localhost/weburl',
    legacyWebUrl: 'http://localhost/legacy',
    studentViewUrl: 'http://localhost/studentview',
    type: 'chapter',
    displayName: 'Any display name',
    children: ['block-children-1', 'block-children-2'],
  },
  'block-children-1': {
    id: 'block-children-1',
    blockId: 'edx_block-1',
    lmsWebUrl: 'http://localhost/weburl',
    legacyWebUrl: 'http://localhost/legacy',
    studentViewUrl: 'http://localhost/studentview',
    type: 'sequential',
    displayName: 'Block children 1',
  },
  'block-children-2': {
    id: 'block-children-2',
    blockId: 'edx_block-2',
    lmsWebUrl: 'http://localhost/weburl',
    legacyWebUrl: 'http://localhost/legacy',
    studentViewUrl: 'http://localhost/studentview',
    type: 'sequential',
    displayName: 'Block children 2',
  },
};

jest.unmock('@edx/frontend-platform/i18n');
jest.unmock('@openedx/paragon');
jest.unmock('@openedx/paragon/icons');

describe('BlocksList Component', () => {
  // eslint-disable-next-line react/prop-types
  const IntlProviderWrapper = ({ children }) => (
    <IntlProvider locale="en">
      {children}
    </IntlProvider>
  );

  let onBlockSelectedMock;

  beforeEach(() => {
    onBlockSelectedMock = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = (overrideProps = {}) => render(
    <IntlProviderWrapper>
      <BlocksList
        blocks={mockBlocks}
        onBlockSelected={onBlockSelectedMock}
        {...overrideProps}
      />
    </IntlProviderWrapper>,
  );

  test('snapshot', async () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('renders without crashing', () => {
    const { getByText } = renderComponent();
    expect(getByText('Any display name')).toBeInTheDocument();
  });

  test('should call onBlockSelected when block name is clicked', () => {
    const { getByTestId } = renderComponent();

    const blockNameButton = getByTestId('block-name');
    fireEvent.click(blockNameButton);
    expect(onBlockSelectedMock).toHaveBeenCalledWith(mockBlocks['block-key']);
  });

  test('should not call onBlockSelected when block navigation is clicked', () => {
    const { getByTestId } = renderComponent();

    const blockNavigateButton = getByTestId('block-navigation');
    fireEvent.click(blockNavigateButton);
    expect(onBlockSelectedMock).not.toHaveBeenCalled();
  });

  test('should show back button when navigation block happens', () => {
    const { getByTestId, getByText } = renderComponent();

    const blockNavigateButton = getByTestId('block-navigation');
    fireEvent.click(blockNavigateButton);

    const backButton = getByTestId('block-back-navigation');
    expect(getByText('Subsections')).toBeInTheDocument();
    expect(getByText('Block children 1')).toBeInTheDocument();
    expect(backButton).toBeInTheDocument();
  });

  test('should show previous block when back navigation button is clicked', () => {
    const { getByTestId, getByText } = renderComponent();

    const blockNavigateButton = getByTestId('block-navigation');
    fireEvent.click(blockNavigateButton);

    const backButton = getByTestId('block-back-navigation');
    expect(getByText('Subsections')).toBeInTheDocument();
    expect(getByText('Block children 1')).toBeInTheDocument();
    expect(backButton).toBeInTheDocument();
    fireEvent.click(backButton);
    expect(getByText('Any display name')).toBeInTheDocument();
  });

  test('should disabled buttons when prop disableBlocks is true', () => {
    const { getByTestId } = renderComponent({ disableBlocks: true });
    const backButton = getByTestId('block-navigation');
    const blockNameButton = getByTestId('block-name');

    expect(backButton).toHaveAttribute('disabled');
    expect(blockNameButton).toHaveAttribute('disabled');
  });
});
