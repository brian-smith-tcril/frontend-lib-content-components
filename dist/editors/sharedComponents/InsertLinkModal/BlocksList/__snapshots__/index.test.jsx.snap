// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlocksList Component snapshot 1`] = `
<div>
  <div
    class="block-list-container"
  >
    <div
      class="pgn-transition-replace-group position-relative"
    >
      <div
        style="padding: .1px 0px;"
      >
        <div
          class="w-100 d-flex justify-content-space-between p-3 pgn__action-row"
        >
          <button
            class="col-11 py-4 btn btn-tertiary"
            data-testid="block-name"
            type="button"
          >
            <span
              class="w-100 text-left"
            >
              Any display name
            </span>
          </button>
          <button
            class="col-1 py-4 btn btn-tertiary"
            data-testid="block-navigation"
            type="button"
          >
            Â 
            <span
              class="pgn__icon btn-icon-after"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m6.49 20.13 1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
